### ISSUE TITLE
Weird issue with symlinks

### ISSUE THREAD

**Issue by cviebrock** [lorisleiva/laravel-deployer#80]
I've noticed that my deployed applications can't find the shared `.env` file.  I think this is because, if I `cd releases/123/` the symlink is:
```
.env -> ../../shared/.env
```
... which resolves.  However `current` is a symlink to `releases/123`, so when I `cd current` I also see:
```
.env -> ../../shared/.env
```
... except that's one too many directories up.  It should be:
```
.env -> ../shared/.env
```

If it helps, my `config/deploy.php` is:
```php
<?php

return [

    'default' => 'upload',

    'strategies' => [
    ],
    
    'hooks' => [
        'start'   => [
        ],

        'build'   => [
        ],

        'ready'   => [
            'artisan:storage:link',
            'artisan:view:clear',
            'artisan:cache:clear',
            'artisan:config:cache',
            'artisan:optimize',
            'artisan:migrate',
            // 'artisan:horizon:terminate',
            'custom:cleanup',
        ],

        'done'    => [
            'fpm:reload',
        ],

        'success' => [
        ],

        'fail'    => [
        ],
    ],

    'options' => [
        'application'     => env('APP_NAME', 'Laravel'),
        'repository'      => 'git@gitlab.com:cviebrock/example.git',
        'branch'          => 'master',
        'http_user'       => 'www-data',
        'php_fpm_service' => 'php7.2-fpm',
        'upload_vendors'  => true,
    ],

    'hosts' => [
        'example.com' => [
            'deploy_path' => '/var/www/example.com',
            'user'        => 'deployer',
            'port'        => 22,
        ],
    ],

    'localhost' => [
    ],

    'include' => [
        'recipes/custom-cleanup.php'
    ],
    'custom_deployer_file' => false,

];
```
And that `custom-cleanup.php` file doesn't have anything to do with symlinks (it just deletes some files that shouldn't be deployed).

Thanks!

**Comment by lorisleiva** [comment]
Hi there 👋 

Can you please tell me how you came up with this hypothesis? Was it always like that or did something change recently? Also do you get an actual error somewhere?

Thanks for reporting this 🙂

**Comment by cviebrock** [comment]
I'll be honest: this is on a side project that I'm very _slowly_ developing, so I haven't looked at the code in a while.  I'm pretty sure it never worked, and I just didn't notice it before.

The way I noticed it was that I have environment settings `HASHID_SALT=foobar` and `HASHID_PADDING=16` which I use with the [/ivanakimov/hashids](https://github.com/ivanakimov/hashids.php) package.  What I was noticing was that locally my hashes weren't matching my production hashes for the same IDs.  A bit of debugging later and I could see that it clearly wasn't reading in those env vars properly in production ... which then led me to notice that the symlinks were "wrong".

There is no actual error that gets thrown.

I'm happy to paste anything else that might help debug this issue.  I did forget to mention that my project's `composer.json` is specifying `"lorisleiva/laravel-deployer": "^0.2.4"`, which is probably a bit out of date by now. 😏

**Comment by lorisleiva** [comment]
Okay thanks for the clarification. Version `0.2.5` only adds support to Laravel 5.7 so nothing that should concern this issue.

So to resume the issue: 
1. You update the `.env` file on the server.
2. You **then** deploy using the `upload` strategy which (from your hooks) clears the config cache and reload FPM.
3. Your application still uses the old environment variables.

If this is a symlink issue, you might want to have a look at this [troubleshooting section](https://github.com/lorisleiva/laravel-deployer/blob/master/docs/troubleshooting.md#my-changes-dont-show-up-after-a-deployment). Let know if this changes anything. 🤞

**Comment by cviebrock** [comment]
It’s not that the .env file can’t be read. It’s the symlink appears to be wrong. 

From `project/releases/123/.env`, you do need to go up two directories to get to `project/shared/.env`.  But from `project/current/.env`, you only need to go up one.

Does that make sense? It seems like following two symlinks (`current/.env` -> `../releases/123/.env` -> `../../shared/.env`) is being interpreted (by the OS?) as just `current/.env` -> `../../shared/.env` ...which is one level up too many.

**Comment by lorisleiva** [comment]
Yeah that’s the problem I had when writing this troubleshooting section, i.e. it was “stuck” on the previous release (not just with the env file though).

If there was something wrong with the way the shared `.env` file is accessed, this would be a fundamental bug and nothing would be working at all. If from the current folder, the relative path to the `.env` file was one level too deep, then you should have a bigger problem than what you are describing.

I’m not sure what else to suggest because unfortunately I cannot know in what ways your application differs from one where this issue doesn’t appear. Be sure to try the troubleshooting section though as this looks like a very similar issue I had with symlinks and OP cache.

**Comment by cviebrock** [comment]
I tried troubleshooting as per those instructions ... no luck.

Long story short, I was using `env('HASHID_SALT')` and `env('HASHID_PADDING')` directly in my code.  I changed that so that I stored it in a `config/medialibrary.php` as:

```php
    'hashid' => [
        'salt' => env('HASHID_SALT'),
        'padding' => env('HASHID_PADDING'),
    ],
```

Then I used `config('medialibrary.hashid.salt')` and `config('medialibrary.hashid.padding')` in my code, and all is fine.

TL;DR: don't try and read environment variables directly; store them in configs first and use that.

**Comment by lorisleiva** [comment]
That one got me a few times too haha. Thanks for letting me know 🙂

**Comment by notflip** [comment]
Is there any real solution to this? It seems my complete .env file is being ignored as well?

**Comment by lorisleiva** [comment]
The issue here was that the Laravel application was using `env()` directly to access its environment variables instead of defining them in the configuration files and then accessing them via `config()`.

When you do that, Laravel cannot cache your config files which is done by default in Laravel Deployer (look at your artisan tasks in your deployment hooks).

This is even written in the official Laravel documentation here: https://laravel.com/docs/5.7/configuration#configuration-caching

> If you execute the config:cache command during your deployment process, you should be sure that you are only calling the env function from within your configuration files. Once the configuration has been cached, the .env file will not be loaded and all calls to the env function will return null.

On the other hand, if the issue seems to be related to symlinks not being updated then I would refer you to the troubleshooting section of the documentation to try out some nginx configurations.
